### PR TITLE
Adding tenant to table mapping API

### PR DESCRIPTION
**Pull request by jenniferdai**
* Currently, it is not easy to get all the tables of a certain tenant. You have to start by getting the instances and manually finding the tables on those instances.
* This API will look at table config in zk and gather the tables that are tagged with the particular tenant.

### CODE REVIEW

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotTenantRestletResource.java [hunk 1]
```diff
@@ -196,6 +200,52 @@ public String listInstanceOrToggleTenantState(
...
+    return getTablesServedFromTenant(tenantName);
+  }
+
+  private String getTablesServedFromTenant(String tenantName)
+      throws JSONException {
+    Set<String> tables = new HashSet<String>();
+    JSONObject resourceGetRet = new JSONObject();
+    Set<String> instances = pinotHelixResourceManager.getAllInstancesForServerTenant(tenantName);
+
+    if (instances.size() == 0) {
```
Comment (mayankshriv) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotTenantRestletResource.java]: Do we enforce and ensure that all created tenants have instances? If not, then is this really an error situation?

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotTenantRestletResource.java [hunk 2]
```diff
@@ -196,6 +200,52 @@ public String listInstanceOrToggleTenantState(
...
+
+    for (String instance : instances) {
+      PropertyKey propertyKey = pinotHelixResourceManager.getKeyBuilder().instance(instance  + CURRENT_STATES_PATH);
+      List<String> childNames = pinotHelixResourceManager.getHelixZkManager().getHelixDataAccessor().getChildNames(propertyKey);
+      if (childNames.size() == 0) {
+        continue;
+      }
+      String child = childNames.get(0);
+      PropertyKey childPropertyKey = pinotHelixResourceManager.getKeyBuilder().instance(instance + CURRENT_STATES_PATH + "/" + child);
+      List<String> tableListZk = pinotHelixResourceManager.getHelixZkManager().getHelixDataAccessor().getChildNames(childPropertyKey);
```
Comment (mayankshriv) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotTenantRestletResource.java]: Is this api expected to return offline and realtime tables separately?

File: pinot-controller/src/test/java/com/linkedin/pinot/controller/api/resources/PinotTenantRestletResourceTest.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    try {
+      // Check that invalid request throws exception
+      new JSONObject(sendGetRequest(_controllerRequestURLBuilder.forTablesFromTenant("randomTenant")));
+      fail("Expected invalid tenant to fail");
+    } catch (Exception e) {
+
+    }
+
+    // Create untagged broker and server instances
```
Comment (mayankshriv) [inline pinot-controller/src/test/java/com/linkedin/pinot/controller/api/resources/PinotTenantRestletResourceTest.java]: Is the section below a copy from elsewhere? If so, convert it into a helper method and reuse.

File: pinot-controller/src/test/java/com/linkedin/pinot/controller/api/resources/PinotTenantRestletResourceTest.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+    }
+
+    // Check that no tables on tenant works
+    tableList = new JSONObject(sendGetRequest(_controllerRequestURLBuilder.forTablesFromTenant("server_REALTIME")));
+    assertEquals(tableList.getJSONArray("tables").length(), 0, "Expected no tables");
+
+    // Try to make sure both kinds of tags work
+    tableList = new JSONObject(sendGetRequest(_controllerRequestURLBuilder.forTablesFromTenant("server")));
+    assertEquals(tableList.getJSONArray("tables").length(), 0, "Expected no tables");
```
Comment (mayankshriv) [inline pinot-controller/src/test/java/com/linkedin/pinot/controller/api/resources/PinotTenantRestletResourceTest.java]: Where's the testcase where we actually get a valid list of tables back? All the tests seem to "Expect no tables".

File: pinot-integration-tests/src/test/java/com/linkedin/pinot/integration/tests/OfflineClusterIntegrationTest.java [hunk 5]
```diff
@@ -40,11 +43,17 @@
 import org.testng.annotations.BeforeClass;
 import org.testng.annotations.Test;
 
+import static org.testng.Assert.*;
```
Comment (mayankshriv) [inline pinot-integration-tests/src/test/java/com/linkedin/pinot/integration/tests/OfflineClusterIntegrationTest.java]: Why static import?

**Comment by kishoreg** [comment]
Can we not get this by looking at tableconfigs? Depending on current state will yield different results depending on state of each node and it’s also more expensive.

File: pinot-controller/src/test/java/com/linkedin/pinot/controller/api/resources/PinotTenantRestletResourceTest.java (hunk 4 above)
Comment (jenniferdai) [reply pinot-controller/src/test/java/com/linkedin/pinot/controller/api/resources/PinotTenantRestletResourceTest.java]: I put the valid one in OfflineClusterIntegrationTest

Comment (mayankshriv) [reply pinot-controller/src/test/java/com/linkedin/pinot/controller/api/resources/PinotTenantRestletResourceTest.java]: Lets put it here as well, to make this test complete.

File: pinot-controller/src/test/java/com/linkedin/pinot/controller/api/resources/PinotTenantRestletResourceTest.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+ * Test for table to tenant mapping. Real working test is inside offline cluster integration test because it requires
+ * segment upload.
+ */
+public class PinotTenantRestletResourceTest extends ControllerTest {
+  private final TableConfig.Builder _offlineBuilder = new TableConfig.Builder(CommonConstants.Helix.TableType.OFFLINE);
+  private static final int NUM_BROKER_INSTANCES = 2;
+  private static final int NUM_SERVER_INSTANCES = 6;
+
+  @BeforeClass
+  public void setUp() {
```
Comment (jackjlli) [inline pinot-controller/src/test/java/com/linkedin/pinot/controller/api/resources/PinotTenantRestletResourceTest.java]: I guess we miss a tearDown class here.

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotTenantRestletResource.java [hunk 7]
```diff
@@ -196,6 +198,42 @@ public String listInstanceOrToggleTenantState(
...
+  private String getTablesServedFromTenant(String tenantName)
+      throws JSONException {
+    Set<String> tables = new HashSet<String>();
+    JSONObject resourceGetRet = new JSONObject();
+
+    for (String table : pinotHelixResourceManager.getAllTables()) {
+      TableConfig tableConfig = pinotHelixResourceManager.getTableConfig(table);
+      String tableConfigTenant = tableConfig.getTenantConfig().getServer();
+      if (tenantName.equals(tableConfigTenant)) {
+        tables.add(table);
```
Comment (jackjlli) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotTenantRestletResource.java]: The logic here shows if there's no segments for this table, this table can still be returned.

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotTenantRestletResource.java [hunk 8]
```diff
@@ -196,6 +198,42 @@ public String listInstanceOrToggleTenantState(
     }
   }
 
+  /**
+   * This method expects a tenant name and will return a list of tables tagged on that tenant. It assumes that the
+   * tagname is for server tenants only. This method will only read CURRENTSTATES in zk.
```
[the remark this answers is not in the document]
Comment (jenniferdai) [reply pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotTenantRestletResource.java]: sure, thanks

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotTenantRestletResource.java (hunk 7 above)
Comment (jenniferdai) [reply pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotTenantRestletResource.java]: I think that's the correct behavior.

**Pull request merged**